README.md
# ESP8266 Cruise Control (GPS/OLED)

[![Build Status](https://travis-ci.com/cohaolain/CruiseControl-ESP8266.svg?branch=master)](https://travis-ci.com/cohaolain/CruiseControl-ESP8266)

This is a project I'm working on so I can have cruise control _(ish)_ in my car.

It uses GPS in order to monitor the speed of your vehicle,
and shows your speed on the OLED display.  
It allows you to set your desired speed limit using two push buttons.  
If you speed, the device screams - the pitch changing with your speeding.

I will write a better description at some point.

## Hardware

-   ESP8266
    -   I used a Heltec WiFi Kit 8
        -   Which the manufacturer & retailer have about 5 completely different pinout diagrams for.
        -   Eventually found a correct one after much digging 😤.
        -   You can find it here: [Heltec Wifi Kit 8 - Pinout Diagram](diagrams/Heltec_WiFi_Kit_8_Pinout_Diagram.pdf)
    -   Not actually using the WiFi capabilities of this myself, _yet_.
-   OLED display
    -   Ideally compatible with `u8g2`.
        The WiFi Kit 8 already has such a display (128x32), quite handy.
        -   It just took a while to figure out which pinout diagram to use 😤.
-   GPS module
    -   I used a module with a NEO-6M.
        -   I _had_ bought a MakerHawk module that claimed to have a NEO-6M.
            It actually turned out to be a NEO-7.  
            I wanted to use the module's EEPROM to store our GPS receiver settings,
            but the newer version doesn't support EEPROM.  
            The NEO-7 module still had the EEPROM soldered on,
            but it was inaccessible. So I'd to return that.  
            Then I got myself a shiny new NEO-6M module!
    -   EEPROM is useful so we
        -   Don't need to connect the RX pin of the GPS module
            (I ran out of easily usable pins on my ESP8266)
        -   Don't need to rely on the BBR (battery-backed RAM)
    -   The default settings are changed so that we
        -   Have a higher navigation refresh rate (5hz compared to 1hz)
        -   Use the automotive navigation profile
        -   Turn off the flashing light (timepulse)
        -   Increase the baud rate (9600 to 19200)
        -   Change what data the module actually provides
    -   Breadboards, pushbuttons, a buzzer, wires, resistors etc.

Will include a schematic in the future.

## Libraries

If you're using the PlatformIO IDE, you can install all
required libraries with the single line:

```bash
platformio lib -g install TinyGPSPlus U8G2
```

You need to have `platformio` binary in your `PATH`. You can do this by adding
the following to the bottom of your `~/.bashrc`:

```bash
export PATH=$PATH:~/.platformio/penv/bin
```

Alternatively, use the relevant library-installation interface in your IDE of choice (e.g. PlatformIO or Arduino).

-   **SoftwareSerial**
    -   For communicating with the GPS module
-   [**TinyGPS++**](https://github.com/mikalhart/TinyGPSPlus)
    -   For dealing with all that GPS data
-   [**u8g2**](https://github.com/olikraus/u8g2/wiki/fntlistall)
    -   For all our display interfacing needs
-   [**ESP8266WiFi**](https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266WiFi)
    -   For (_now_), this is just to turn off the WiFi capabilities of the ESP8266.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2023-07-20
Haven't worked on this in years, but just wanted to clean it up!

### Added

-   A startup screen/logo thing. While we search for a GPS signal!

### Changed

-   Seperated Buzzer functions into an included file.

## [0.2.1] - 2020-05-29

### Added

-   The **`safetyMultiplier`**, which adds 5% to the current speed.
    This will lead to earlier warnings, and just slower driving in general.
-   More debugging serial output. Now in the form of NMEA messages.

### Changed

-   Major overhaul of GPS serial data handling.
-   Removed frame which surrounds the UI on the OLED.
-   Current list of speed limits.
-   Logic for storing new GPS readings.
-   Flipped the OLED orientation.
-   Debug serial baud rate is now 1843200
    (not useful to link it to the GPS baud rate anymore).
-   Doesn't keep buzzing if GPS signal is lost while
    speeding is detected.
-   The u8g2 library class instance is now an global variable,
    instead of manually passing a reference to it in constructors.
-   Travis builds are now initially published as drafts.
-   Moving average now only considers past 3 readings (750ms), not 10.

### Fixed

-   Incorrect declaration of number of possible speed limits
    no longer results in accessing beyond the end of an array
    when changing the current limit.

## [0.1.3] - 2020-05-19

### Added

-   New speedometer component to UI.
-   Better documentation for installing dependency libraries.
-   Warning tone now stops if we don't have reliable up-to-date GPS data.
-   There is now a cap on the max pitch of the warning buzzer.

### Changed

-   GPS data is now sent to serial for debugging purposes.
-   All baud rates are now 9600.
-   U-Blox Neo-6M configuration
    -   Update rate is now 4 Hz instead of 5 Hz.
-   Low-speed GPS data is no longer sometimes discarded
-   GPS data now must have a HDOP of less than 4, instead of 50.
-   Adjusted text alignment on invalid-fix page.

### Fixed

-   U-Blox Neo-6M configuration
    -   Now provides the correct messages.
-   Invalid fix screen now shown if the speed data given is invalid.
-   GPS smoothing resulting in new values not being stored.

### Removed

-   U-Blox Neo-7 configuration, as I can't test this for now.

## [0.1.2] - 2020-04-24

### Added

-   Checks for missing/stale serial data/GPS updates
-   Better representation of the GPS status on display
-   Pinout diagram for Heltec WiFi Kit 8
-   Build status badge for README.md

### Changed

-   Button actions now occur before and after GPS serial read,
    not just before.
-   Now can yield to h/w interrupts when waiting for serial data

## [0.1.1] - 2020-04-21

### Added

-   Travis CI integration
    -   This will check that tagged project versions' firmwares compile
        succesfully, and then add the resulting firmware image as a release on
        GitHub.

## [0.1.0] - 2020-04-21

### Added

-   First version of the project source
-   Some documentation


lib/StartLogo/StartLogo.h
#pragma once

#include <U8g2lib.h>

extern U8G2 globalInstanceU8G2;

#define start_logo_width 128
#define start_logo_height 32
// Sorry, I really can't draw.
static unsigned char start_logo_bits[] U8X8_PROGMEM = {
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xe2, 0x00, 0x20, 0x00, 0x00, 0x18,
    0x80, 0x00, 0x80, 0x00, 0x00, 0x00, 0x00, 0x30, 0x00, 0x00, 0x32, 0x00,
    0x20, 0x04, 0x00, 0x0e, 0x80, 0x03, 0x80, 0x00, 0x00, 0x00, 0x00, 0xee,
    0x01, 0x00, 0x19, 0x38, 0x05, 0xe3, 0x01, 0x01, 0xe0, 0xc0, 0x9c, 0x00,
    0x00, 0x00, 0x80, 0x3f, 0x0f, 0x00, 0x09, 0x0c, 0x25, 0x21, 0x83, 0x80,
    0x9d, 0xb8, 0x96, 0x00, 0x00, 0x00, 0xe0, 0xfc, 0x18, 0x00, 0x19, 0x0c,
    0x25, 0xf3, 0x81, 0xc0, 0x9d, 0x18, 0x91, 0x00, 0x00, 0x00, 0xf8, 0x3f,
    0x30, 0x00, 0xf1, 0x09, 0x27, 0x62, 0x80, 0x40, 0xad, 0x08, 0x99, 0x00,
    0x00, 0x00, 0x0c, 0x63, 0xa0, 0x81, 0x00, 0x00, 0xa0, 0xe3, 0x01, 0xd3,
    0x25, 0x0f, 0x8e, 0x03, 0x00, 0x00, 0x04, 0x42, 0xfc, 0x8f, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x1c, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x03, 0xc2,
    0xc7, 0x8a, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x80, 0x01, 0x73, 0xee, 0x4d, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0xc0, 0x01, 0xfd, 0x96, 0x4e, 0x10, 0x20,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xfe, 0x1f, 0xc7,
    0xad, 0x4c, 0xf0, 0x29, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x80, 0xbb, 0xfe, 0xcf, 0x55, 0x2c, 0x30, 0x69, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0xe0, 0xfe, 0xd7, 0xd6, 0x65, 0x26, 0xa0, 0x59,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x30, 0xff, 0x7d, 0xd5,
    0xe5, 0x21, 0xe0, 0x60, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0xb0, 0xef, 0x3f, 0x53, 0xfd, 0x13, 0x00, 0x22, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0xd8, 0xd3, 0x0f, 0x73, 0xff, 0x13, 0x00, 0x1c,
    0x00, 0x00, 0x38, 0x00, 0x00, 0x00, 0x00, 0x00, 0x08, 0x7f, 0x04, 0xfd,
    0xf9, 0x13, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x88, 0x25, 0x04, 0x1d, 0xb8, 0x09, 0x00, 0x00, 0x00, 0x80, 0x03, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x90, 0x29, 0xe4, 0x1f, 0xc0, 0x08, 0x1c, 0x00,
    0x00, 0x00, 0x00, 0x02, 0x03, 0x04, 0x06, 0x00, 0xf0, 0x29, 0xfc, 0x1f,
    0x00, 0x08, 0x03, 0x00, 0x18, 0x00, 0x00, 0x02, 0x03, 0xc4, 0x11, 0x00,
    0x00, 0xff, 0xcf, 0x1f, 0x00, 0x88, 0x10, 0x00, 0x06, 0x00, 0x1e, 0x02,
    0x07, 0x44, 0x00, 0x00, 0x00, 0xc0, 0x81, 0x0f, 0x00, 0xc4, 0x00, 0xe0,
    0x00, 0x00, 0x31, 0x0a, 0x05, 0x04, 0x00, 0x07, 0x00, 0x00, 0x00, 0x03,
    0x00, 0x44, 0xd0, 0x33, 0xdc, 0x87, 0x20, 0x9e, 0x44, 0x84, 0x27, 0x0f,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x44, 0x50, 0x12, 0xd6, 0x84, 0x20, 0x96,
    0xe7, 0xc4, 0x26, 0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0xc2, 0x53, 0x33,
    0x9a, 0x04, 0x13, 0x52, 0xa8, 0x44, 0x26, 0x09, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x02, 0x80, 0x27, 0xee, 0x04, 0x0e, 0x42, 0xa8, 0x5c, 0x2f, 0x09,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00,
    0x40, 0xc0, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};

class StartLogo
{
public:
    StartLogo(int x_pos, int y_pos);
    void displayLogo(int duration);

private:
    U8G2 u8g2 = globalInstanceU8G2;
    int x;
    int y;
};


lib/Speedometer/Speedometer.cpp
#include <Speedometer.h>

Speedometer::Speedometer(int height, int width, int x, int initLimit)
{
    updateRange(initLimit);

    radius = (4 * pow(height, 2) + pow(width, 2)) / (8 * height);
    center_x = x + width / 2;
    center_y = u8g2.getDisplayHeight() + (radius - height);
}

void Speedometer::draw(double speed)
{
    u8g2.drawCircle(center_x, center_y, radius);

    drawTick(limit);
    drawNeedle(speed);
}

void Speedometer::updateRange(int newLimit)
{
    limit = newLimit;
    lowerBound = limit * 0.8;
}

double mapf(double x, double in_min, double in_max, double out_min, double out_max)
{
    const double delta = x - in_min;
    const double out_range = out_max - out_min;
    const double in_range = in_max - in_min;

    return delta * out_range / in_range + out_min;
}

void Speedometer::drawTick(double speed, double shadeFrom, double shadeTo)
{
    double theta = mapf(speed, lowerBound, limit, normal_range, PI - normal_range);

    theta = min(theta, PI + overflow_range);
    theta = max(theta, -overflow_range);

    const int inner_x = center_x - (cos(theta) * (radius * shadeFrom));
    const int inner_y = center_y - (sin(theta) * (radius * shadeFrom));
    const int outer_x = center_x - (cos(theta) * (radius * shadeTo));
    const int outer_y = center_y - (sin(theta) * (radius * shadeTo));

    u8g2.drawLine(inner_x, inner_y, outer_x, outer_y);
}

void Speedometer::drawTick(double speed)
{
    drawTick(speed, 0.7, 1);
}

void Speedometer::drawNeedle(double speed)
{
    drawTick(speed, 0, 1);
}


src/cruise_control.cpp
// Bare essentials
#include <SoftwareSerial.h>
#include <ESP8266WiFi.h>

// GPS & display libraries
#include <TinyGPS++.h>
#include <U8g2lib.h>

// My libraries
#include <Speedometer.h>
#include <StartLogo.h>
#include <Buzzer.h>

#define DEBUG

const int buzzerPin = D8;
const int greenPin = D7;
const int redPin = D6;

const int numPrevReadings = 3;
const int numSpeeds = 11;
const int speeds[numSpeeds] = {5, 10, 15, 20, 25, 30, 50, 60, 80, 100, 120};
const uint8_t *fonts[] = {u8g2_font_profont12_tf, u8g2_font_profont17_mf, u8g2_font_profont12_mf};

const double safetyMultiplier = 1.05;

bool wasSpeeding;
bool deviceDisabled;

int currentSpeedLimIndex = 6;
double prevReadings[numPrevReadings];

bool pendingButtonAction;
bool greenPushed;
bool redPushed;

int lastUpdateGPS = 0;
int lastButtonPress = 0;
int lastSerialRead = 0;

TinyGPSPlus gps;
SoftwareSerial ss(14, -1);
U8G2_SSD1306_128X32_UNIVISION_F_HW_I2C u8g2(U8G2_R0, /* reset=*/16, SCL, SDA);
U8G2 globalInstanceU8G2 = u8g2;

Speedometer speedometer(24, 32, 16, speeds[currentSpeedLimIndex]);
StartLogo Logo(0, 0);

void preinit()
{
    // Don't need WiFi for this project, turn it off to save power
    WiFi.preinitWiFiOff();
}

double currentSpeed()
{
    double acc = 0;
    for (double val : prevReadings)
    {
        acc += val;
    }
    return (acc / numPrevReadings) * safetyMultiplier;
}

bool updateSpeedLim(int indexDelta)
{
    noTone(buzzerPin);
    if (currentSpeedLimIndex + indexDelta >= 0 && currentSpeedLimIndex + indexDelta < numSpeeds)
    {
        currentSpeedLimIndex += indexDelta;
        speedometer.updateRange(speeds[currentSpeedLimIndex]);
        return true;
    }
    return false;
}

// Interrupt for button presses
ICACHE_RAM_ATTR void onButtonPressed()
{
    if (lastButtonPress + 150 <= millis())
    {
        greenPushed = digitalRead(greenPin);
        redPushed = digitalRead(redPin);
        pendingButtonAction = true;
    }
    lastButtonPress = millis();
}

void performPendingButtonActions()
{
    if (pendingButtonAction)
    {
        if (greenPushed && redPushed)
        {
            deviceDisabled ^= 1;
            u8g2.setPowerSave(deviceDisabled);
            actionBeep(!deviceDisabled);
        }
        else if ((greenPushed || redPushed))
        {
            ackBeep(!updateSpeedLim(greenPushed ? 1 : -1));
        }
        pendingButtonAction = false;
    }
}

void addReading(double reading)
{
    double temp1 = prevReadings[0];
    for (int i = 1; i < numPrevReadings; i++)
    {
        double temp2 = prevReadings[i];
        prevReadings[i] = temp1;
        temp1 = temp2;
    }
    prevReadings[0] = reading;
}

void updateGPS()
{
    if (gps.speed.isUpdated())
    {
        double reading = gps.speed.kmph();
        if (reading >= 2)
            addReading(reading);
        else
            addReading(0);
        lastUpdateGPS = millis();
    }
}

bool waitForSerial()
{
    performPendingButtonActions();

    if (ss.available())
    {
        while (ss.available())
        {
            String nmeaString = ss.readStringUntil('\n');
            for (char c : nmeaString)
            {
                gps.encode(c);
            }
            Serial.println(nmeaString);
        }

        updateGPS();

        Serial.printf("$GPTXT,CC says,hdop (%s): %.2f\n", gps.hdop.isValid() ? "valid" : "invalid", gps.hdop.hdop());
        Serial.printf("$GPTXT,CC says,location (%s): %f, %f\n", gps.location.isValid() ? "valid" : "invalid", gps.location.lat(), gps.location.lng());
        Serial.printf("$GPTXT,CC says,speed (%s): %.2f km/h\n", gps.speed.isValid() ? "valid" : "invalid", gps.speed.kmph());
        Serial.printf("$GPTXT,CC says,stated speed (%s): %.2f km/h\n", gps.speed.isValid() ? "valid" : "invalid", currentSpeed());
        Serial.printf("$GPTXT,CC says,passed: %d, failed: %d, success: %.2f\%\n", gps.passedChecksum(), gps.failedChecksum(), ((double)gps.passedChecksum()) / max((uint32_t)1, (gps.passedChecksum() + gps.failedChecksum())));
        Serial.print("$GPTXT,CC says,prevReadings: ");
        for (auto val : prevReadings)
            Serial.printf("%.2f ", val);
        Serial.println("\n");

        lastSerialRead = millis();
    }

    return millis() - lastSerialRead <= 1000;
}

static char *formattedTime()
{
    char buf[16];
    TinyGPSTime t = gps.time;
    if (t.isValid())
    {
        sprintf(buf, "%02d:%02d", t.hour(), t.minute());
    }
    else
    {
        sprintf(buf, "??:??");
    }
    return buf;
}

static char *formattedSpeed()
{
    char buf[32];
    char padBuf[32];
    dtostrf(currentSpeed(), 3, 1, buf);
    sprintf(padBuf, "%5s", buf);
    return padBuf;
}

bool allPrevReadingsZero()
{
    for (double val : prevReadings)
    {
        if (val != 0)
            return false;
    }
    return true;
}

void printAccuracy(int x, int y)
{
    u8g2.setFont(fonts[0]);
    u8g2.setCursor(x, y);
    if (gps.hdop.isValid())
    {
        u8g2.print(gps.hdop.hdop());
    }
    else
    {
        u8g2.print("...");
    }
}

void printTime()
{
    u8g2.setFont(fonts[2]);
    u8g2.drawStr(128 - 30, 8, formattedTime());
}

void printSpeed()
{
    u8g2.setFont(fonts[1]);
    u8g2.drawStr(49, 15, formattedSpeed());
    u8g2.setFont(fonts[0]);
    u8g2.setCursor(70, 28);
    u8g2.print("km/h");
}

void printLimit(int x, int y)
{
    char buf[64];
    sprintf(buf, "%d", speeds[currentSpeedLimIndex]);
    u8g2.setDrawColor(1);
    u8g2.drawDisc(x, y - 1, 10, U8G2_DRAW_ALL);
    u8g2.setDrawColor(0);
    u8g2.drawStr(x + 1 - (6 * ((int)log10(speeds[currentSpeedLimIndex]) + 1)) / 2, y + 3, buf);
    u8g2.setDrawColor(1);
}

void tooFast()
{
    double speedingAmount = ((currentSpeed() / speeds[currentSpeedLimIndex]) - 1) * 100;
    int freq = min(map(speedingAmount, 0, 50, 220, 800), 800l);
    tone(buzzerPin, freq);
    wasSpeeding = true;
}

void checkSpeed()
{
    if (currentSpeed() <= speeds[currentSpeedLimIndex] && wasSpeeding)
    {
        wasSpeeding = !wasSpeeding;
        noTone(buzzerPin);
    }
    else if (currentSpeed() > speeds[currentSpeedLimIndex])
        tooFast();
}

void showNoSerial()
{
    u8g2.clearBuffer();
    u8g2.setFont(fonts[1]);
    u8g2.setCursor(6, 12);
    u8g2.print("No GPS Module");
    u8g2.setCursor(12, 28);
    u8g2.print("Attached :(");
    u8g2.sendBuffer();
}

void showInvalid()
{
    u8g2.clearBuffer();
    u8g2.setFont(fonts[1]);
    u8g2.setCursor(22, 12);
    u8g2.print("Searching");
    u8g2.setCursor(18, 28);
    u8g2.print("for GPS...");
    u8g2.sendBuffer();
}

void setup()
{
    // Serial for debugging
    Serial.begin(1843200);
    Serial.println("Begin!");

    // Software serial for GPS
    ss.begin(9600, SWSERIAL_8N1, 14, -1, false, 256);
    ss.setTimeout(10);

    // Setup OLED display
    u8g2.begin();

    // Setup buttons
    pinMode(redPin, OUTPUT);
    pinMode(greenPin, OUTPUT);
    attachInterrupt(digitalPinToInterrupt(redPin), onButtonPressed, RISING);
    attachInterrupt(digitalPinToInterrupt(greenPin), onButtonPressed, RISING);

    // Setup UI
    speedometer.updateRange(speeds[currentSpeedLimIndex]);

    // Show logo
    Logo.displayLogo(0);
    u8g2.sendBuffer();
    delay(10000);
}

void loop()
{
    if (waitForSerial())
    {
        if (gps.hdop.hdop() >= 4 || !gps.speed.isValid())
        {
            if (wasSpeeding)
                noTone(buzzerPin);
            showInvalid();
        }
        else
        {
            checkSpeed();

            // Render UI
            u8g2.clearBuffer();

            printAccuracy(2, 10);
            printSpeed();
            printLimit(112, 17);
            speedometer.draw(currentSpeed());

            // Write UI into display buffer
            u8g2.sendBuffer();
        }
    }
    else
    {
        if (wasSpeeding)
            noTone(buzzerPin);
        showNoSerial();
    }
}
